### PR TITLE
Ingest COG terms, pathways and functions

### LINKED ISSUE #1333: backend ingest updates to support functional search by COG terms
related to FY25 Q1 milestone https://github.com/microbiomedata/issues/issues/522

Assumes records would be in the functional_annotation_agg collection with a prefix of 'COG'
example 
{
  "metagenome_annotation_id": "nmdc:wfmgan-11-ndgg7v31.1",
  "gene_function_id": "COG:COG0001",
  "count": 56
}

2020 release term names, pathways and categories on nersc /global/cfs/cdirs/m3408/functional_annotation_mappings/cog2020

There will be similar tickets for Pfam and GO so we should take the opportunity to generalize https://github.com/microbiomedata/nmdc-server/blob/main/nmdc_server/ingest/kegg.py

Acceptance criteria: ingest process accepts COG terms such that they appear in the postgres database. There will be a separate ticket for data portal front end changes to make the functional search more generic.

### PR DESCRIPTION
**Pull request by naglepuff**
Fix #1333

### Changes
In additional to populating postgres tables with IDs, definitions of, and relationships between KEGG modules, pathways, and orthologs, ingest now populates those tables with COG terms, functions and pathways. The files used to ingest this information are found in the NERSC CFS. From the NMDC home directory (`/global/cfs/cdirs/m3408/`) they can be found under `/ingest/cog/`.

New models and an associated migration have also been included. The `ko_term_to_text` table has been overloaded to include a mapping of ID to description for COG terms/pathways/functions as well. Note that COG pathways don't have a unique identifier associated with them, so the description is used for both the `term` and `text` column.

Additionally, documents in the mongo collection `functional_annotation_agg` whose `gene_function_id` begins with `COG` will be ingested.

### To test
To test locally you must modify the setup you normally use for local ingest to make sure the necessary files are mounted. See `ingest/kegg.py` for some additional details. Additionally, you must make sure that there are some representative documents in your `functional_annotation_agg` collection. After running local ingest, inspect the relevant tables to make sure that COG terms, functions and pathways are ingested, along with the `functional_annotation_agg` documents.

### Deployment Notes
I've rearranged the files in NERSC so that all of the files that ingest needs are in `.../m3408/ingest/`. I think we should update our mount points to mount to a subdirectory of `.../m3408` instead of mounting to the whole entire project directory.

### CODE REVIEW

**Comment by ssarrafan** [comment]
@marySalvi can this be reviewed by tomorrow so we can close the issue this sprint?

**Comment by marySalvi** [comment]
> @marySalvi can this be reviewed by tomorrow so we can close the issue this sprint?

Yes, I'll have this reviewed today

File: nmdc_server/ingest/kegg.py [hunk 1]
```diff
@@ -44,18 +119,20 @@ def ingest_tree(node: dict) -> None:
...
+            fieldnames=keys.get("fieldnames", None),
+        )
     return records
 
 
 def ingest_ko_module_map(db: Session) -> None:
+    """Ingest a mapping of KEGG modules to terms and COG functions to terms."""
     db.execute(f"truncate table {KoTermToModule.__tablename__}")
 
     datafile = Path(__file__).parent / "data" / "kegg_module_ko_terms.tab.txt"
```
Comment (marySalvi) [inline nmdc_server/ingest/kegg.py:138]:
```suggestion
    datafile = Path(__file__).parent / "data" / "kegg" / "kegg_module_ko_terms.tab.txt"
```

File: nmdc_server/ingest/kegg.py [hunk 2]
```diff
@@ -66,8 +143,25 @@ def ingest_ko_module_map(db: Session) -> None:
...
+            [CogTermToFunction(term=record[0], function=record[1]) for record in records]
+        )
+        db.commit()
+
 
 def ingest_ko_pathway_map(db: Session) -> None:
+    """Ingest a mapping of KEGG pathways to terms and COG pathways to COG terms."""
     db.execute(f"truncate table {KoTermToPathway.__tablename__}")
 
     datafile = Path(__file__).parent / "data" / "ko_term_pathways.tab.txt"
```
Comment (marySalvi) [inline nmdc_server/ingest/kegg.py:167]:
```suggestion
    datafile = Path(__file__).parent / "data" / "kegg" / "ko_term_pathways.tab.txt"
```

**Review by marySalvi — CHANGES_REQUESTED** [review]
Looks good with some minor changes. I also had to run inside container. I don't know if that it a blocker

File: nmdc_server/ingest/kegg.py [hunk 3]
```diff
@@ -66,8 +143,25 @@ def ingest_ko_module_map(db: Session) -> None:
         )
         db.commit()
 
+    with open(COG_TERM_DEFS) as fd:
```
Comment (marySalvi) [inline nmdc_server/ingest/kegg.py:146]:
```suggestion
    cogterms = Path(__file__).parent / "data" / "cog" / "cog-20.def.tab"
    with open(cogterms) as fd:
```

this is the docker culprit

File: nmdc_server/ingest/kegg.py [hunk 4]
```diff
@@ -77,3 +171,11 @@ def ingest_ko_pathway_map(db: Session) -> None:
             [KoTermToPathway(term=row["KO_id"], pathway=row["image_id"]) for row in reader]
         )
         db.commit()
+
+    with open(COG_TERM_DEFS) as fd:
```
Comment (marySalvi) [inline nmdc_server/ingest/kegg.py:175]:
```suggestion
    cogterms = Path(__file__).parent / "data" / "cog" / "cog-20.def.tab"
    with open(cogterms) as fd:
```
ditto from above

**Review by marySalvi — CHANGES_REQUESTED** [review]

**Comment by naglepuff** [comment]
@ssarrafan we're going to hold off on merging this until next week. There are some aspects of this part of ingest that we'd like to discuss a little bit more internally. I expect that to be resolved early next week though (Monday or Tuesday)

**Review by marySalvi — APPROVED** [review]

**Pull request merged**